### PR TITLE
WIP: Add unit test for static_library() with no source files

### PR DESCRIPTION
**Pull request by xclaesse**
This only add a unit test to try to reproduce #3735 on the CI.

### CODE REVIEW

**Comment by nirbheek** [comment]
`lib.exe` has the same issue, so we should solve it in a cross-platform agnostic way.

**Comment by jon-turney** [comment]
See also #2876

**Comment by xclaesse** [comment]
I have no idea how to fix this, any one has suggestions?

**Comment by nirbheek** [comment]
> I have no idea how to fix this, any one has suggestions?

There are two ways:

1. If the target has no sources, use the objects directly.
a) For libraries built by us, convert the `link_whole:` argument to actually refer to the objects inside that library. We actually might want to always do this.
b) For libraries that aren't built by us, extract them into the builddir and use the objects directly.

2. If the target has no sources, create an empty source file and add it as a source.

**Comment by xclaesse** [comment]
>     1. If the target has no sources, use the objects directly.
>        a) For libraries built by us, convert the `link_whole:` argument to actually refer to the objects inside that library. We actually might want to always do this.

I believe we do that already now, so this issue could be already fixed in the case there is at least one static library we build ourself.

>        b) For libraries that aren't built by us, extract them into the builddir and use the objects directly.

We don't do that, but a library that is only concat of system static libraries is really corner case, we could get away with keeping that bug. But ideally it would be nice to extract .a archives indeed...

>     2. If the target has no sources, create an empty source file and add it as a source.

That seems an easy fix indeed, I suspect some projects already work around it like that themself.

**Comment by jon-turney** [comment]
>  If the target has no sources, create an empty source file and add it as a source.

It's not immediately clear to me that an empty file is a valid source file for all the languages we support (any of which might be the only language we have a compiler for).

At the least, we should emit a portability warning when a llibrary has an empty sources list.

**Comment by eli-schwartz** [comment]
This originally did get checked, but in commit d070100c8c0529db984b4cb10de745e762b9199a that got broken and later fixed in commit 5ca5e0c900340110eb26bcf4491033e9d1bacf08.

tl;dr We should not be allowing this at all. build targets of all kinds should have either objects (maybe from link_whole) or source files to compile, but if we don't then we emit a deprecation warning for now.
